gmt-4.5.14/src/x2sys/x2sys_init.1
.TH X2SYS_INIT 1 "1 Nov 2015" "GMT 4.5.14" "Generic Mapping Tools"
.SH NAME
x2sys_init \- Initialize x2sys data base for track data files
.SH SYNOPSIS
\fBx2sys_init\fP \fITAG\fP \fB\-D\fP\fIdeffile\fP [ \fB\-Cc\fP|\fBf\fP|\fBg\fP|\fBe\fP ] [ \fB\-E\fP\fIsuffix\fP ] 
[ \fB\-F\fP ] [ \fB\-Gd\fP|\fBg\fP ] [ \fB\-I\fP\fIdx\fP[/\fIdy\fP] ] [ \fB\-Nd\fP|\fBs\fP\fIunit\fP ] 
[ \fB\-R\fP\fIwest\fP/\fIeast\fP/\fIsouth\fP/\fInorth\fP[\fBr\fP] ] [ \fB\-V\fP ] [ \fB\-Wt\fP|\fBd\fP\fIgap\fP ] [ \fB\-m\fP[\fBi\fP|\fBo\fP][\fIflag\fP] ]
.SH DESCRIPTION
\fBx2sys_init\fP is the starting point for anyone wishing to use x2sys; it initializes
a set of data bases that are particular to one kind of track data.  These data,
their associated data bases, and key parameters are given a short-hand
notation called an x2sys TAG.   The TAG keeps track of settings such as file format,
whether the data are geographic or not, and the binning resolution for track indices.
Running \fBx2sys_init\fP is a prerequisite to running any of the other x2sys programs,
such as \fBx2sys_binlist\fP, which will
create a crude representation of where each data track go within the domain
and which observations are available; this information serves as input to
\fBx2sys_put\fP which updates the track data base. Then, \fBx2sys_get\fP can
be used to find which tracks and data are available inside a given region.
With that list of tracks you can use \fBx2sys_cross\fP to calculate track
crossovers, use \fBx2sys_report\fP to report crossover statistics or
\fBx2sys_list\fP to pull out selected crossover information
that \fBx2sys_solve\fP can use to determine track-specific systematic corrections.  These
corrections may be used with \fBx2sys_datalist\fP to extract corrected
data values for use in subsequent work.
.TP
\fITAG\fP
The unique name of this data type x2sys TAG.
.TP
\fB\-C\fP
Select procedure for along-track distance calculation when needed by other programs:
.br
\fBc\fP Cartesian distances [Default, unless \fB\-G\fP is set].
.br
\fBf\fP Flat Earth distances.
.br
\fBg\fP Great circle distances [Default if \fB\-G\fP is set].
.br
\fBe\fP Geodesic distances on current \fBGMT\fP ellipsoid.
.TP
\fB\-D\fP
Definition file prefix for this data set [See DEFINITION FILES below for more information].
Specify full path if the file is not in the current directory.
.SH OPTIONS
No space between the option flag and the associated arguments.
.TP
\fB\-E\fP
Specifies the file extension (suffix) for these data files.  If not given we use the
definition file prefix as the suffix (see \fB\-D\fP).
.TP
\fB\-F\fP
Force creating new files if old ones are present [Default will abort if old TAG files are found].
.TP
\fB\-G\fP
Selects geographical coordinates.  Append \fBd\fP for discontinuity at the
Dateline (makes longitude go from -180 to + 180) or \fBg\fP for discontinuity
at Greenwich (makes longitude go from 0 to 360 [Default]).  If not given we
assume the data are Cartesian.
.TP
\fB\-I\fP
\fIx_inc\fP [and optionally \fIy_inc\fP] is the grid spacing. Append \fBm\fP to
indicate minutes or \fBc\fP to indicate seconds for geographic data.  These
spacings refer to the binning used in the track bin-index data base.
.TP
\fB\-m\fP
Multiple segment file(s).  Segments are separated by a special record.
For ASCII files the first character must be \fIflag\fP [Default is '>'].
For binary files all fields must be NaN and \fB\-b\fP must
set the number of output columns explicitly.  By default the \fB\-m\fP
setting applies to both input and output.  Use \fB\-mi\fP and \fB\-mo\fP
to give separate settings to input and output.
.TP
\fB\-N\fP
Sets the units used for distance and speed when requested by other programs.
Append \fBd\fP for distance or \fBs\fP for speed, then give the desired \fIunit\fP as
\fBc\fP (Cartesian userdist or userdist/usertime), \fBe\fP (meter or m/s), \fBk\fP (km or km/hr), \fBm\fP (miles or miles/hr), or
\fBn\fP (nautical miles or knots).  [Default is \fB\-Ndk\fP \fB\-Nse\fP (km and m/s) if \fB\-G\fP is set
and \fB\-Ndc\fP and \fB\-Nsc\fP otherwise (Cartesian units)].
.TP
\fB\-R\fP
\fIwest, east, south,\fP and \fInorth\fP specify the Region of interest, and you may specify them
in decimal degrees or in [+-]dd:mm[:ss.xxx][W|E|S|N] format.  Append \fBr\fP if lower left and upper right
map coordinates are given instead of w/e/s/n.  The two shorthands \fB\-Rg\fP and \fB\-Rd\fP stand for global domain
(0/360 and -180/+180 in longitude respectively, with -90/+90 in latitude). Alternatively, specify the name
of an existing grid file and the \fB\-R\fP settings (and grid spacing, if applicable) are copied from the grid.
For Cartesian data just give \fIxin/xmax/ymin/ymax\fP.  This sets the complete domain
for the relevant track data set.
.TP
\fB\-V\fP
Selects verbose mode, which will send progress reports to stderr [Default runs "silently"].
.TP
\fB\-W\fP
Give \fBt\fP or \fBd\fP and append the corresponding
maximum time gap (in user units; this is typically seconds [Infinity]), or distance
(for units, see \fB\-N\fP) gap [Infinity]) allowed between
the two data points immediately on either side of a crossover.  If these limits are
exceeded then a data gap is assumed and no COE will be determined.
.SH DEFINITION FILES
These *.def files contain information about the data file format and
have two sections: (1) header information and (2) column information.  All
header information starts with the character # in the first column, immediately
followed by an upper-case directive.  If the directive takes an argument it is
separated by white-space.  You may append a trailing # comments.
Five directives are recognized:
.br
.sp
\fBASCII\fP states that the data files are in ASCII format.
.br
\fBBINARY\fP states that the data files are native binary files.
.br
\fBNETCDF\fP states that the data files are COARDS-compliant 1-D netCDF files.
.br
\fBSKIP\fP takes an integer argument which is either the number of lines to
skip (when reading ASCII files) or the number of bytes to skip (when reading
native binary files). Not used with netCDF files.
.br
\fBGEO\fP indicates that these files are geographic data sets, with periodicities in
the \fIx\fP-coordinate (longitudes).  Alternatively, use \fB\-G\fP.
.br
\fBMULTISEG\fP means each track consists of multiple segments separated
by a \fBGMT\fP multisegment header (alternatively, use \fB\-m\fP when defining
the system TAG). Not used with netCDF files.
.br
.sp
The column information consists of one line per column in the order the
columns appear in the data file.  For each column you must provide  seven
attributes:
.sp
.br
\fIname type NaN NaN-proxy scale offset oformat\fP
.br
.sp
\fIname\fP is the name of the column variable.  It is expected that you will use
the special names \fIlon\fP (or \fIx\fP if Cartesian) and \fIlat\fP (or \fIy\fP)
for the two required coordinate columns, and \fItime\fP when optional time data
are present.
.br
\fItype\fP is always \fBa\fP for ASCII representations of numbers, whereas for
binary files you may choose among
\fBc\fP for signed 1-byte character (-127,+128), \fBu\fP for unsigned
byte (0-255), \fBh\fP for signed 2-byte integers (-32768,+32767),
\fBi\fP for  signed 4-byte integers (-2,147,483,648,+2,147,483,647),
\fBf\fP for 4-byte floating points and \fBd\fP for 8-byte double
precision floating points.  For netCDF, simply use \fBd\fP as netCDF
will automatically handle type-conversions during reading.
.br
\fINaN\fP is Y if certain values (e.g, -9999) are to be replaced by NAN, and N otherwise.
.br
\fINaN-proxy\fP is that special value (e.g., -9999).
.br
\fIscale\fP is used to multiply the data after reading.
.br
\fIoffset\fP is used to add to the scaled data.
.br
\fIoformat\fP is a C-style format string used to print values from this column.
.br
.sp
If you give - as the \fIoformat\fP then \fBGMT\fP's formatting machinery will be used\"'
instead (i.e., \fBD_FORMAT\fP, \fBPLOT_DEGREE_FORMAT\fP, \fBPLOT_DATE_FORMAT\fP, \fBPLOT_CLOCK_FORMAT\fP).
Some file formats already have definition files premade.  These include
mgd77 (for plain ASCII MGD77 data files), mgd77+ (for enhanced MGD77+
netCDF files), gmt (for old mgg supplement binary files), xy (for plain ASCII
x, y tables), xyz (same, with one z-column), geo (for plain ASCII longitude,
latitude files), and geoz (same, with one z-column).
.SH EXAMPLES
If you have a large set of track data files you can organize them using the
x2sys tools.  Here we will outline the steps.  Let us assume that your track
data file format consist of 2 header records with text information followed
by any number of identically formatted data records with 6 columns (lat,
lon, time, obs1, obs2, obs3) and that files are called *.trk. We will call
this the "line" format.  First, we create the line.def file:
.br
.sp
\fB\fP# Define file for the line format
.br
\fB\fP#ASCII		# File is ASCII
.br
\fB\fP#SKIP 2		# Skip 2 header records
.br
\fB\fP#GEO		# Data are geographic
.br
\fB\fP#name	type 	NaN	NaN-proxy	scale offset oformat
.br
lat	a	N	0	1	0	%9.5f
.br
lon	a	N	0	1	0	%10.5f
.br
time	a	N	0	1	0	%7.1f
.br
obs1	a	N	0	1	0	%7.2f
.br
obs2	a	N	0	1	0	%7.2f
.br
obs3	a	N	0	1	0	%7.2f
.br
.sp
.br
Next we create the TAG and the TAG directory with the databases for these line track files.
Assuming these contain geographic data and that we want to keep track of the
data distribution at a 1 x 1 degree resolution, with distances in km calculated
along geodesics and with speeds given in knots, we may run
.br
.sp
\fBx2sys_init\fP LINE \fB\-V\fP \fB\-G\fP \fB\-D\fP line \fB\-Rg\fP \fB\-Ce\fP \fB\-Ndk\fP \fB\-NsN\fP \fB\-I\fP 1/1 \fB\-E\fP trk
.br
.sp
where we have selected LINE to be our x2sys tag.  When x2sys tools try to read your
line data files they will first look in the current directory and second look
in the file \fITAG\fP_paths.txt for a list of additional directories to examine.
Therefore, create such a file (here LINE_paths.txt) and stick the full paths to
your data directories there.  All TAG-related files (definition files, tag files,
and track data bases created) will be expected to be in the directory pointed to
by \fB$X2SYS_HOME\fP/\fITAG\fP (in our case \fB$X2SYS_HOME\fP/LINE).  Note that
the argument to \fB\-D\fP must contain the full path if the *.def file is not in
the current directory.  \fBx2sys_init\fP will copy this file to the \fB$X2SYS_HOME\fP/\fITAG\fP
directory where all other x2sys tools will expect to find it.
.br
.TP
.B Create tbf file(s):
Once the (empty) TAG databases have been initialized we go through a
two-step process to populate them.  First we run \fBx2sys_binlist\fP on
all our track files to create one (or more) multi-segment track bin-index
files (tbf).  These contain information on which 1 x 1 degree bins (or any
other blocksize; see \fB\-I\fP) each
track has visited and which observations (in your case obs1, obs2, obs3)
were actually observed (not all tracks may have all three kinds of
observations everywhere).  For instance, if your tracks are listed in
the file tracks.lis we may run this command:
.br
.sp
\fBx2sys_binlist\fP \fB\-V\fP \fB\-T\fP LINE :tracks.lis > tracks.tbf
.TP
.B Update index data base:
Next, the track bin-index files are fed to \fBx2sys_put\fP
which will insert the information into the TAG databases:
.br
.sp
\fBx2sys_put\fP \fB\-V\fP \fB\-T\fP LINE tracks.tbf
.TP
.B Search for data:

gmt-4.5.14/ChangeLog
-----------------------------------------------------------------------------------------
	$Id: ChangeLog 10331 2015-10-29 19:41:56Z pwessel $
-----------------------------------------------------------------------------------------

-----------------------------------------------------------------------------------------
2015-11-01	pwessel
	* GMT 4.5.14 finalized and released
-----------------------------------------------------------------------------------------
2015-10-29	pwessel
	* meca/psmeca.c:	Enforce rakes and strikes to be in -180/+180 range.
				Issue #799, for gmt4 only as not a problem in gmt5.

2015-07-28	pwessel
	* meca/psmeca.c:	Be able to read any value x and y coordinates

2015-06-24	pwessel
	* gmt_support.c:	Fix bug in jump near Dateline in gmt_inonout_sphpol_count,
				affecting inside/outside spherical cases, reported in message-2219

2015-03-18	pwessel
	* share/cpt/GMT_wysiwyg.cpt:	Fix typo in blue for high end [Issue #689].

2015-01-16	pwessel
	* gmt_plot.c:		Fix issue # 662 with graph arrows for negative scales.

2015-01-10	pwessel
	* gmt_plot.c:		Flipping the n/s sides array when lon and -y axes
				fixes issue #520

2015-01-01	pwessel
	* configure.ac:		Moved to 4.5.14 and back to svn release only.
-----------------------------------------------------------------------------------------
2015-01-01	pwessel
	* GMT 4.5.13 finalized and released [GSHHG 2.3.4 was released on Jan 1 as well]
-----------------------------------------------------------------------------------------
2014-12-28	pwessel
	* gmt_map.c:		Needed separate GMT_cartesian_overlap for non-geographic
				projections.  Fixes issue # 657.

2014-09-29	pwessel
	* gmt_plot.c:		Left a PSL variable undefined when grdimage plotted a grid
				with x = longitude and y = Cartesian.

2014-09-09	pwessel
	* grdfilter.c:		Fix y-shifts when -I sets smaller output.  Issue #616.
	* gmt_bcr.c:		Fix problem with points slightly outside causing an
				access violation; see issue #620.

2014-08-29	pwessel
	* sample1d.c:		Fix resampling for decreasing t values.  Issue #608.

2014-08-05	pwessel
	* gmt_map.c:		Could run into trouble clipping polygons for non-periodic
				maps.  Added more checks in GMT_wesn_clip.

2014-07-31	pwessel
	* gmt_map.c:		Rect clipping with -R<w/s/e/n>r could introduce stray lines
				when dataset has features about 180 degrees away from
				the user's area. See issue # 594.

2014-07-30	pwessel
	* configure.ac:		Set default GSHHG version to 2.3.2 (released August 1, 2014)
	* psxy[z].c:		Fixed wrap-around issue in fault lines.

2014-07-27	pwessel
	* gmtselect.c:		The -Z option needed to pass records with NaNs.

2014-07-19	pwessel
	* x2sys/x2sys_solve.c:	Fix problems that arise when a track has less crossovers
				than model parameters.
	* gmtmath.c:		Fix issue #123 with files and -C.

2014-07-01	pwessel
	* configure.ac:		Set default GSHHG version to 2.3.1 (released July 1, 2014)
	
2014-06-28	pwessel
	* mgd77/mgd77sniffer.c:	Needed to check that grid values used in decimate function
	 			also were within reasonable range.

2014-05-26	pwessel
	* gmt_map.c:		The map region clip path for -JE/e was hardwired to a full
				circle and did not care about -R setting.
				When -R...r and -Je/E was given the search for enclosing
				geographic boundaries failed when either the S or N pole
				was the projection's antipole.
2014-05-25	pwessel
	* grdmath.c:		Fix issue #566 bug in grdmath's KM2DEG operator.

2014-03-19	pwessel
	* gmtselect.c:		The -Fpolygon option might miss points because the io-machinery
				would sometimes set the min/max longitudes found incorrectly.

2014-03-14	pwessel
	* gmt_shore.c:		Give more error feedback if nc_open fails due to old library.

2014-03-01	pwessel
	* configure.ac:		Moved to 4.5.13 and back to svn release only.

-----------------------------------------------------------------------------------------
2014-03-01	pwessel
	* GMT 4.5.12 finalized and released [GSHHG 2.3.0 was released on Feb 1]
-----------------------------------------------------------------------------------------
2014-02-17	fwobbe
	* gmt_init.c etc.:	File locking -- if enabled via -DFLOCK -- is
				non-compulsory now.  This means if fcntl fails for some reason,
				GMT continues to access the file.  Hence, it is recommended to
				configure GMT with --enable-flock even if fcntl might be buggy
				on some ancient NFS shares.

2014-02-17	pwessel
	* gmt_map.c:		Size for memory allocation cast the product of nx * ny to
				size_t but overflow had already happen as calculation was
				done using ints, then cast to size_t. Must cast nx and ny
				separately.
2014-02-10	jluis
	* misc/gmt2kml.c:	Make sure the default style IDs are unique for each process.

2014-02-07	jluis
	* ps2raster.c:		Make sure that images in KML files do not carry a directory name

2014-02-03	pwessel
	* pslib.c:		Auto-widen paragraph width to at least fit widest word.

2014-02-01	pwessel
	* configure.ac:		Since GSHHG 2.3.0 was released today, make that the default

2014-01-26	pwessel
	* gmt_grdio.c:		Building the grid command string could exceed its length if
				the argument became exactly the max length of the array; then
				a final extra character \0 would be added and exceed array length.
				Thanks to Joachim Saul for pointing it out.

2014-01-22	pwessel
	* pscoast.c:		The processing of -I -N levels and pens could get out of sync
				so that the wrong pen was used for the specified feature.

2014-01-16	remko
	* pslib.c:		Take superscripts and subscripts into account when determining
				dimensions of text box.

2014-01-10	pwessel
	* grdmath.c:		Had to let SDIST return km instead of degrees since it cannot

gmt-4.5.14/src/x2sys/x2sys_init.c
/*-----------------------------------------------------------------
 *	$Id: x2sys_init.c 10289 2014-12-28 21:17:06Z pwessel $
 *
 *      Copyright (c) 1999-2015 by P. Wessel
 *      See LICENSE.TXT file for copying and redistribution conditions.
 *
 *      This program is free software; you can redistribute it and/or modify
 *      it under the terms of the GNU General Public License as published by
 *      the Free Software Foundation; version 2 or any later version.
 *
 *      This program is distributed in the hope that it will be useful,
 *      but WITHOUT ANY WARRANTY; without even the implied warranty of
 *      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *      GNU General Public License for more details.
 *
 *      Contact info: gmt.soest.hawaii.edu
 *--------------------------------------------------------------------*/
/* x2sys_init will accept command line options to specify track data
 * file formats, region of interest, bin spacing, etc, and a unique
 * system identifier.  It will then initialize the databases associated
 * with this system where information about the tracks and their
 * intersections will be stored.
 *
 * Author:	Paul Wessel
 * Date:	14-JUN-2004
 * Version:	1.1, based on the spirit of the old mgg/s_system code
 *
 */

#include "x2sys.h"
extern void x2sys_set_home (void);

int main (int argc, char **argv)
{
	char *sfile = CNULL, *TAG = CNULL;
	time_t right_now;
#ifndef WIN32
	struct passwd *pw = NULL;
#endif
	char tag_file[BUFSIZ], track_file[BUFSIZ], bin_file[BUFSIZ], def_file[BUFSIZ];
	char path_file[BUFSIZ], path[BUFSIZ], line[BUFSIZ];

	GMT_LONG error = FALSE, force = FALSE;

	FILE *fp = NULL, *fp_def = NULL;

	int n_tags = 0, i, n_found = 0, r_entry, i_entry, d_entry, d_start;
	int e_entry,  g_entry, m_entry, wd_entry, wt_entry, n_entry[2], c_entry, n_n = 0;

	double x_min = 0.0, x_max = 0.0, y_min = 0.0, y_max = 0.0, bin_x = 0.0, bin_y = 0.0;

	c_entry = n_entry[0] = n_entry[1] = r_entry = i_entry = d_entry = e_entry = g_entry = m_entry = wd_entry = wt_entry = 0;

	argc = (int)GMT_begin (argc, argv);
	
	for (i = (int)strlen(argv[0]); i >= 0 && argv[0][i] != '/'; i--);
	X2SYS_program = &argv[0][i+1];	/* Name without full path */

	for (i = 1; i < argc; i++) {
		if (argv[i][0] == '-') {
			switch (argv[i][1]) {
				/* Common parameters */

				case 'R':
					r_entry = i;
				case 'H':
				case 'V':
				case '\0':
					error += GMT_parse_common_options (argv[i], &x_min, &x_max, &y_min, &y_max);
					break;

				/* Supplemental parameters */

				case 'C':	/* Distance calculation flag */
					if (!strchr ("cefg", (int)argv[i][2])) {
						fprintf(stderr, "%s: ERROR -C: Flag must be c, f, g, or e\n", GMT_program);
						error++;
					}
					c_entry = i;
					break;
				case 'D':
					sfile = &argv[i][2];
					d_entry = i;
					break;
				case 'E':
					e_entry = i;
					break;
				case 'G':	/* Geographical coordinates, set discontinuity */
					g_entry = i;
					break;
				case 'F':
					force = TRUE;
					break;
				case 'I':
					if (argv[i][2]) GMT_getinc (&argv[i][2], &bin_x, &bin_y);
					i_entry = i;
					break;
				case 'M':
				case 'm':
					m_entry = i;
					break;
				case 'N':	/* Distance and speed unit selection */
					switch (argv[i][2]) {
						case 'd':	/* Distance unit selection */
						case 's':	/* Speed unit selection */
							if (!strchr ("cekmn", (int)argv[i][3])) {
								fprintf(stderr, "%s: ERROR -N%c: Unit must be c, e, k, m, or n\n", GMT_program, argv[i][2]);
								error++;
							}
							break;
						default:
							fprintf(stderr, "%s: ERROR -N: Choose from -Nd and -Ns\n", GMT_program);
							error++;
							break;
					}
					if (n_n >= 2) {
						fprintf(stderr, "%s: ERROR -N: Can only be given once each for -Nd -Ns\n", GMT_program);
						error++;
					}
					else
						n_entry[n_n++] = i;
					break;
				case 'W':
					switch (argv[i][2]) {
						case 't':	/* Get new timegap */
							wt_entry = i;
							break;
						case 'd':	/* Get new distgap */
							wd_entry = i;
							break;
						default:
							fprintf (stderr, "%s: Syntax Error: -Wt|d<width>\n", GMT_program);
							error++;
							break;
					}
					break;
				default:
					error = TRUE;
					break;
			}
		}
		else {
			TAG = argv[i];
			n_tags++;
		}
	}

	if (argc == 1 || error || GMT_give_synopsis_and_exit) {
		fprintf (stderr, "x2sys_init %s - Initialize a new track system database\n\n", X2SYS_VERSION);
		fprintf (stderr, "usage: x2sys_init <TAG> [-Cc|f|g|e] [-D<deffile>] [-E<suffix>] [-F] [-G[d/g]] [-I[<binsize>]]\n");
		fprintf (stderr, "\t[-N[d|s][c|e|k|m|n]]] [%s] [-V] [-Wt|d|n<gap>] [%s]\n\n", GMT_Rgeo_OPT, GMT_m_OPT);
		fprintf (stderr, "\t<TAG> is the unique system identifier.  Files created will be placed in\n");
		fprintf (stderr, "\t   the directory X2SYS_HOME/<TAG>.\n");

		if (GMT_give_synopsis_and_exit) exit (EXIT_FAILURE);

		fprintf (stderr, "\n\tOPTIONS:\n");
		fprintf (stderr, "\t-C Select procedure for along-track distance and azimuth calculations:\n");
		fprintf (stderr, "\t   c Plain Cartesian\n");
		fprintf (stderr, "\t   f Flat Earth\n");
		fprintf (stderr, "\t   g Great circle [Default]\n");
		fprintf (stderr, "\t   e Ellipsoidal (geodesic) using current ellipsoid\n");
		fprintf (stderr, "\t-D definition file for the track data set [<TAG>.def]\n");
		fprintf (stderr, "\t-E Extension (suffix) for these data files\n");
		fprintf (stderr, "\t   [Default equals the prefix for the definition file]\n");
		fprintf (stderr, "\t-F Force creating new files if old ones are present [Default will abort if old files are found]\n");
		fprintf (stderr, "\t-G for geographical coordinates.  Append g for discontinuity at Greenwich (output 0/360 [Default])\n");
		fprintf (stderr, "\t   and append d for discontinuity at Dateline (output -180/+180)\n");
		fprintf (stderr, "\t-I sets bin size for track bin index output [1/1]\n");
		fprintf (stderr, "\t-N Append (d)istances or (s)peed, and your choice for unit. Choose among:\n");
		fprintf (stderr, "\t   c Cartesian distance (user-dist-units, user user-dist-units/user-time-units)\n");
		fprintf (stderr, "\t   e Metric units I (meters, m/s)\n");
		fprintf (stderr, "\t   k Metric units II (km, km/hr)\n");
		fprintf (stderr, "\t   m British/US units (miles, miles/hr)\n");
		fprintf (stderr, "\t   n Nautical units (nautical miles, knots)\n");
		fprintf (stderr, "\t   [Default is -Ndk -Nse]\n");
		GMT_explain_option ('R');
		fprintf (stderr, "\t   [Default region is 0/360/-90/90]\n");
		GMT_explain_option ('V');
		fprintf (stderr, "\t-W Set maximum gaps allowed at crossover.  Option may be repeated.\n");
		fprintf (stderr, "\t   -Wt sets maximum time gap (in user units) [Default is infinite]\n");
		fprintf (stderr, "\t   -Wd sets maximum distance gap (in user units) [Default is infinite]\n");
		GMT_explain_option ('m');
		exit (EXIT_FAILURE);
	}

	if (n_tags == 0) {
		fprintf (stderr, "%s: No system tag given!\n", GMT_program);
		exit (EXIT_FAILURE);
	}
	if (n_tags > 1) {
		fprintf (stderr, "%s: Only give one system tag!\n", GMT_program);
		exit (EXIT_FAILURE);
	}
	if (r_entry && (x_min >= x_max || y_min >= y_max)) {
		fprintf (stderr, "%s: -R given inconsistent values!\n", GMT_program);
		exit (EXIT_FAILURE);
	}
	if (i_entry && (bin_x <= 0.0 || bin_y <= 0.0)) {
		fprintf (stderr, "%s: -Idx/dy must be positive!\n", GMT_program);
		exit (EXIT_FAILURE);
	}
	if (!sfile) sfile = TAG;	/* Default def file */
	sprintf (def_file, "%s.def", sfile);
	if (access (def_file, R_OK)) {	/* No such local *.def file */
		fprintf (stderr,"%s: Unable to find local definition file : %s\n", GMT_program, def_file);
		exit (EXIT_FAILURE);
	}
	else if ((fp_def = fopen (def_file, "r")) == NULL) {
		fprintf (stderr,"%s: Unable to open local definition file : %s\n", GMT_program, def_file);
		exit (EXIT_FAILURE);
	}
	for (d_start = (int)strlen (sfile)-1; d_start >= 0 && sfile[d_start] != '/'; d_start--);	/* Find pos of last slash */
	d_start++;		/* Find start of file name */
	
	/* Determine the TAG directory */
	
	x2sys_set_home ();
	x2sys_path (TAG, path);
	if (x2sys_access (TAG, R_OK)) {	/* No such dir */
		if (mkdir (path, (mode_t)0777)) {
			fprintf (stderr,"%s: Unable to create TAG directory : %s\n", GMT_program, path);
			exit (EXIT_FAILURE);
		}
	}
	else if (!force) {	/* Directory exists but -F not on */
		fprintf (stderr,"%s: TAG directory already exists: %s\n", GMT_program, path);
		exit (EXIT_FAILURE);
	}
	
	/* Initialize the system TAG files in X2SYS_HOME/TAG */

	sprintf (tag_file, "%s/%s.tag", TAG, TAG);
	sprintf (def_file, "%s/%s.def", TAG, &sfile[d_start]);
	sprintf (path_file, "%s/%s_paths.txt", TAG, TAG);
	sprintf (track_file, "%s/%s_tracks.d", TAG, TAG);
	sprintf (bin_file, "%s/%s_index.b", TAG, TAG);

	if (!x2sys_access (tag_file, R_OK)) {
		fprintf (stderr,"%s: File exists: %s\n", GMT_program, tag_file);
		x2sys_path (tag_file, path);
		if (force) {
			if (remove (path))
				fprintf (stderr,"%s: Unable to remove %s\n", GMT_program, path);
			else
				fprintf (stderr,"%s: Removed file %s\n", GMT_program, path);
		}
		else 
			n_found++;
	}
	if (!x2sys_access (def_file, R_OK)) {
		fprintf (stderr,"%s: File exists: %s\n", GMT_program, def_file);
		x2sys_path (def_file, path);
		if (force) {
			if (remove (path))
				fprintf (stderr,"%s: Unable to remove %s\n", GMT_program, path);
			else
				fprintf (stderr,"%s: Removed file %s\n", GMT_program, path);
		}
		else 
			n_found++;
	}
	if (!x2sys_access (track_file, R_OK)) {
		fprintf (stderr,"%s: File exists: %s\n", GMT_program, track_file);
		x2sys_path (track_file, path);
		if (force) {
			if (remove (path))
				fprintf (stderr,"%s: Unable to remove %s\n", GMT_program, path);
			else
				fprintf (stderr,"%s: Removed file %s\n", GMT_program, path);
		}
		else 
			n_found++;
	}
	if (!x2sys_access (path_file, R_OK)) {
		fprintf (stderr,"%s: File exists: %s\n", GMT_program, path_file);
		x2sys_path (path_file, path);
		if (force) {
			if (remove (path))
				fprintf (stderr,"%s: Unable to remove %s\n", GMT_program, path);
			else
				fprintf (stderr,"%s: Removed file %s\n", GMT_program, path);
		}
		else 
			n_found++;
	}
	if (!x2sys_access (bin_file, R_OK)) {
		fprintf (stderr,"%s: File exists: %s\n", GMT_program, bin_file);
		x2sys_path (bin_file, path);
		if (force) {
			if (remove (path))
				fprintf (stderr,"%s: Unable to remove %s\n", GMT_program, path);
			else
				fprintf (stderr,"%s: Removed file %s\n", GMT_program, path);
		}
		else 
			n_found++;
	}
	if (n_found) {
		fprintf (stderr,"%s: Remove/rename old files or use -F to overwrite\n", GMT_program);
		exit (EXIT_FAILURE);
	}


	if (gmtdefs.verbose) fprintf (stderr, "%s: Initialize %s\n", GMT_program, tag_file);
	if ((fp = x2sys_fopen (tag_file, "w")) == NULL) {
		fprintf (stderr,"%s: Could not create file %s\n", GMT_program, tag_file);
		exit (EXIT_FAILURE);
	}

        right_now = time ((time_t *)0);
	fprintf (fp, "# TAG file for system: %s\n", TAG);
	fprintf (fp, "#\n# Initialized on: %s", ctime (&right_now));
#ifndef WIN32
	if ((pw = getpwuid (getuid ())) != NULL)
		fprintf (fp, "# Initialized by: %s\n#\n", pw->pw_name);
	else
#endif
		fprintf (fp, "# Initialized by: unknown\n#\n");
	fprintf (fp, "-D%s", &sfile[d_start]);	/* Now a local *.def file in the TAG directory */
	if (c_entry) fprintf (fp, " %s", argv[c_entry]);
	if (e_entry) fprintf (fp, " %s", argv[e_entry]);
	if (g_entry) fprintf (fp, " %s", argv[g_entry]);
	if (m_entry) fprintf (fp, " -m%s", &argv[m_entry][2]);
	if (n_entry[0]) fprintf (fp, " %s", argv[n_entry[0]]);
	if (n_entry[1]) fprintf (fp, " %s", argv[n_entry[1]]);
	if (wt_entry) fprintf (fp, " %s", argv[wt_entry]);
	if (wd_entry) fprintf (fp, " %s", argv[wd_entry]);
	(i_entry) ? fprintf (fp, " %s", argv[i_entry]) : fprintf (fp, " -I1/1");
	(r_entry) ? fprintf (fp, " %s", argv[r_entry]) : fprintf (fp, " -R0/360/-90/90");
	fprintf (fp, "\n");
	x2sys_err_fail (x2sys_fclose (tag_file, fp), tag_file);

	/* Initialize the system's definition file  */

	if (gmtdefs.verbose) fprintf (stderr, "%s: Initialize %s\n", GMT_program, def_file);
	if ((fp = x2sys_fopen (def_file, "w")) == NULL) {
		fprintf (stderr, "%s: Could not create %s\n", GMT_program, def_file);
		exit (EXIT_FAILURE);
	}
	while (fgets (line, BUFSIZ, fp_def)) fprintf (fp, "%s", line);
	x2sys_err_fail (x2sys_fclose (def_file, fp), def_file);
	fclose (fp_def);	/* Close local def file */

	/* Initialize the system's tracks data base  */

	if (gmtdefs.verbose) fprintf (stderr, "%s: Initialize %s\n", GMT_program, track_file);
	if ((fp = x2sys_fopen (track_file, "w")) == NULL) {
		fprintf (stderr, "%s: Could not create %s\n", GMT_program, track_file);
		exit (EXIT_FAILURE);
	}
	x2sys_err_fail (x2sys_fclose (track_file, fp), track_file);

	/* Initialize the system's index data base  */

	if (gmtdefs.verbose) fprintf (stderr, "%s: Initialize %s\n", GMT_program, bin_file);
	if ((fp = x2sys_fopen (bin_file, "wb")) == NULL) {
		fprintf (stderr,"%s: Could not create %s\n", GMT_program, bin_file);
		exit (EXIT_FAILURE);
	}
	x2sys_err_fail (x2sys_fclose (bin_file, fp), bin_file);

	/* Initialize the system's track path file  */

	if (gmtdefs.verbose) fprintf (stderr, "%s: Initialize %s\n", GMT_program, path_file);
	if ((fp = x2sys_fopen (path_file, "wb")) == NULL) {
		fprintf (stderr,"%s: Could not create %s\n", GMT_program, path_file);
		exit (EXIT_FAILURE);
	}
	fprintf (fp, "# Directories with data files for TAG %s\n", TAG);
	fprintf (fp, "# The current directory is always searched first.\n");
	fprintf (fp, "# Add full paths to search additional directories\n");
	x2sys_err_fail (x2sys_fclose (path_file, fp), path_file);
	
	if (gmtdefs.verbose) fprintf (stderr, "%s completed successfully\n", GMT_program);

	GMT_free ((void *)X2SYS_HOME);
	GMT_end (argc, argv);

	exit (EXIT_SUCCESS);
}
